docs/structured-streaming-programming-guide.md

```diff
@@ -1647,6 +1648,26 @@ For example, sorting on the input stream is not supported, as it requires keepin
 track of all the data received in the stream. This is therefore fundamentally hard to execute 
 efficiently.
 
+### Limitation of global watermark
+
+In Append mode, some stateful operations could emit rows older than current watermark plus allowed late record delay,
+which are "late rows" in downstream stateful operations (as Spark uses global watermark) and these rows can be discarded.
+This is a limitation of global watermark, and it could bring correctness issue.
+
+Spark will check the logical plan of query and log warning when Spark detects such pattern.
```

Comment:
log warning -> log a warning
such pattern -> such a pattern

But above don't you say that Spark can't check this?

Reply:
Above sentence means "Spark can't check the state function is implemented via semantics of output mode" because we don't restrict anything on implementing state function (even FlatMapGroupsWithState knows about output mode but doesn't do anything with output mode), whereas this is checking logical plan to find the pattern which is possible - assuming the state function is implemented via semantics of output mode.